docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```diff
@@ -0,0 +1,99 @@
+# ADR 059: Evidence Composition and Lifecycle
+
+## Changelog
+
+- 04/09/2020: Initial Draft
+
+## Scope
+
+This document is a draft designed to collate together and surface some predicaments involving evidence in Tendermint with the aim of trying to form a more harmonious
+composition which hopefully shall be later captured in one or two ADR's. The scope does not extend to the verification nor detection of certain types of evidence but concerns itself mainly with the general form of evidence and its lifecycle.
+
+## Background
+
+For a long time `DuplicateVoteEvidence`, formed in the consensus engine, was the only evidence Tendermint had. It was produced whenever two votes from the same validator in the same round
+was observed and thus it was made in a per validator way. It was predicted that there may come more forms of evidence and thus `DuplicateVoteEvidence` was used as the model for the `Evidence` interface and also for the form of the evidence data sent to the application. It is important to note that Tendermint concerns itself just with the detection and reporting of evidence and it is the responsibility of the application to exercise punishment.
+
+Tendermint has now introduced a new type of evidence to protect light clients from being attacked. This `LightClientAttackEvidence` is vastly different to `DuplicateVoteEvidence` in that it is physically a much different size containing a complete signed header and validator set. It is formed within the light client, not the consensus engine and requires a lot more information from state to verify. Finally it batches validators together as opposed to having individual evidence. This evidence stretches the existing mould that was used to accomodate new types of evidence and has thus caused us to reconsider how evidence should be formatted and processed.
+
+## Possible Approaches
+
+### Individual framework
+
+Evidence remains on a per validator basis. This causes the least disruption to the current processes but requires that we break `LightClientAttackEvidence` into several pieces of evidence for each malicious validator. This not only has performance consequences in that there are n times as many database operations and that the gossiping of evidence will require more bandwidth then necessary (by requiring a header for each piece) but it potentially impacts our ability to validate it. In batch form, the full node can run the same process the light client did to see that 1/3 validating power was present in both the common block and the conflicting block whereas this becomes more difficult to verify individually. Not only that, but `LightClientAttackEvidence` also deals with amnesia attacks which unfortunately have the characteristic where we know the set of validators involved but not the subset that were actually malicious (more to be said about this later).
+
+### Hybrid Framework
```

Comment:
another hybrid approach could be 

a) send / receive evidence in batches (duplicate votes, light client attacks, etc.)
b) break evidence per validator 
c) store also per validator & height

Reply:
and how would evidence be committed on chain? individual or batch

Reply:
on evidence is in blocks, so it's already "batched" in this way.

Reply:
Sure but you could have two different data structures, a batch of individual evidence or a single batched evidence, where the latter is likely to be smaller in size

Reply:
sure